File: Zeze/Util/IdentityHashMap.cs
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Runtime.CompilerServices;

namespace Zeze.Util
{
    public sealed class IdentityHashMap<K, V> : ConcurrentDictionary<K, V>
    {
        private class IdentityEqualityComparer : IEqualityComparer<K>
        {
            bool IEqualityComparer<K>.Equals(K x, K y)
            {
                return object.ReferenceEquals(x, y);
            }

            int IEqualityComparer<K>.GetHashCode(K obj)
            {
                return RuntimeHelpers.GetHashCode(obj);
            }
        }
        private static readonly IdentityEqualityComparer comparer = new IdentityEqualityComparer();

        public IdentityHashMap() : base(comparer)
        { 
        }
    }
}

File: ZezeJava/src/Zeze/Util/IdentityHashMap.java
package Zeze.Util;

import Zeze.*;

public final class IdentityHashMap<K, V> extends java.util.concurrent.ConcurrentHashMap<K, V> {
	private static class IdentityEqualityComparer implements IEqualityComparer<K> {
		public final boolean equals(K x, K y) {
			return x == y;
		}

		public final int hashCode(K obj) {
			return RuntimeHelpers.hashCode(obj);
		}
	}
	private static IdentityEqualityComparer comparer = new IdentityEqualityComparer();

	public IdentityHashMap() {
		super(comparer);
	}
}